component/common/loki/client/queue_client.go

```diff
@@ -283,16 +285,26 @@ func (c *queueClient) AppendEntries(entries wal.RefEntries, segment int) error {
 	c.seriesLock.RLock()
 	l, ok := c.series[entries.Ref]
 	c.seriesLock.RUnlock()
+	var maxSeenTimestamp int64 = -1
 	if ok {
 		for _, e := range entries.Entries {
 			c.appendSingleEntry(segment, l, e)
+			if e.Timestamp.Unix() > maxSeenTimestamp {
+				maxSeenTimestamp = e.Timestamp.Unix()
+			}
 		}
 		// count all enqueued appended entries as received from WAL
 		c.markerHandler.UpdateReceivedData(segment, len(entries.Entries))
 	} else {
 		// TODO(thepalbi): Add metric here
 		level.Debug(c.logger).Log("msg", "series for entry not found")
 	}
+
+	// emit metric tracking max seen timestamp in WAL entry
+	if maxSeenTimestamp != -1 {
+		c.qcMetrics.lastReadTimestamp.WithLabelValues().Set(float64(maxSeenTimestamp))
+	}
```

Comment:
nit: won't this _always_ be the case when there's at least 1 entry?

Reply:
Yes! Upon an `AppendEntries` call there will always be at least one entry, if not it wouldn't have been decoded. Removing if.